ArmPkg/Library/ArmArchTimerLib/ArmArchTimerLib.c

```diff
@@ -130,6 +131,15 @@ MicroSecondDelay (
 
   TimerTicks64 += SystemCounterVal;
 
+  if (TimerTicks64 < SystemCounterVal) {
+    // Wrap around - wait for wrap
+    PreviousSystemCounterVal = SystemCounterVal;
+    while (PreviousSystemCounterVal <= SystemCounterVal) {
+      // Still going up
+      PreviousSystemCounterVal = SystemCounterVal;
+      SystemCounterVal = ArmGenericTimerGetSystemCount ();
+    }
+  }

```

Comment:
OK, so as I understand it, the failure mode is that once in a blue moon, we will return from this function too early, right? Wast his found by inspection?

The handling looks correct to me, but given that this is effectively dead code, I was hoping that instead, we might tweak the original logic so it behaves correctly even when the counter wraps.

If not, please expand the comment to clarify that you are looping waiting for the counter to wrap, so that the subsequent conditional will not prematurely conclude that the work is done.

Reply:
I found this by inspection while I was wondering why my timeouts were totally wrong vs. the docs  when running on my FVP (they were taking many times longer than they should). It immediately stood out to me that it didn't handle wrap-around while trying to diagnose why. I've dealt with wraparounds with timestamps before to see the problem immediately. Rough back of a napkin math will have this happen every 40-something days or so given a 50mhz timer clock which is what I had, so not that rare.

It literally does what it says on the tin. It handles it (correctly) in the case of a wrap. This is the smallest tweaking of logic I can think of by just handling the wrap part with a separate pre-loop. 

So if not ... ummm what? If it isn't simple tweaking of the logic? (To me it is). Can you suggest something simpler? If it's not dead code? I don't think it's dead. It's a rare bug but still there. The comment that says "Wrap around - wait for wrap" is not saying enough? That in the case of wrap around, wait for the wrap (as it's inside the if block)? I don't know what to add to the comment that's says what you ask me to say that isn't already there? If I added "IF wrap around, THEN wait for wrap" to make it clear the comment is inside the if? Just checking that that is what you want as I might be missing something... ?